llvm/lib/Transforms/Coroutines/CoroFrame.cpp

```diff
@@ -1614,6 +1621,24 @@ struct AllocaUseVisitor : PtrUseVisitor<AllocaUseVisitor> {
       // suspend point between lifetime markers. This should also cover the
       // case of a single lifetime.start intrinsic in a loop with suspend point.
       if (PI.isEscaped()) {
+        // If there is no explicit lifetime.end, then assume the address can
```

Comment:
Due to github review system, I can only put the comments here for the above code. According to our discussion, we can only check the whether the lifetime.{start, end} cross suspend point. We don't need to check users. Then it can be cheaper.

Reply:
This algorithm as described would not work.

The problem is that checking for `lifetime.start` -> `coro.suspend` -> `lifetime.end` doesn't cover the case where `lifetime.end` is missing, which is the original bug.

Even if we modify the algorithm to check for `lifetime.end`, `unreachable`, and `return`, we would still need to make sure there is no `lifetime.end` between the `lifetime.start` and `coro.suspend` (since that would mean the object can live on the stack). The simpler algorithm therefore is to just check for `lifetime.start` -> (no `lifetime.end`) -> `coro.suspend`.

Reply:
I think I misunderstood your initial comment - my apologies.

If I understand correctly, you're saying that the following code before the line `if (PI.isEscaped()) {`:

```cpp
for (auto *I : Users)
  for (auto *S : LifetimeStarts)
    if (Checker.isDefinitionAcrossSuspend(*S, I))
      return true;
```

is not necessary because we're checking for all `Users` instead of just `llvm.lifetime.end`, so therefore, we can replace those lines of code with the `lifetime.end` logic I wrote below.

I made this change and `ninja check-llvm` passed locally, so LMK if this is what you meant.

Reply:
Yes, we don't need check users if we check `lifetime.end`.